src/Cuco.API/Controllers/UserController.cs

```diff
@@ -0,0 +1,101 @@
+using Cuco.Application.Providers;
+    using Cuco.Commons.Base;
+using Cuco.Domain.Roles.Models.Consts;
+using Cuco.Domain.Roles.Services.Repositories;
+using Cuco.Domain.Users.Models.DTO;
+using Cuco.Domain.Users.Models.Entities;
+using Cuco.Domain.Users.Services.Repositories;
+using Microsoft.AspNetCore.Authorization;
+using Microsoft.AspNetCore.Mvc;
+
+namespace Cuco.API.Controllers;
+
+[ApiController]
+[Produces("application/json")]
+[Route("api/user")]
+public class UserController : ControllerBase
+{
+    private readonly IUnitOfWork _unitOfWork;
+    private readonly IUserProvider _userProvider;
+    private readonly IUserRepository _userRepository;
+
+    public UserController(IUserProvider userProvider, IUserRepository userRepository, IUnitOfWork unitOfWork)
+    {
+        _userProvider = userProvider;
+        _userRepository = userRepository;
+        _unitOfWork = unitOfWork;
+    }
+
+    [HttpPost]
+    [Authorize(Roles = $"{RoleNames.Admin}")]
+    [ProducesResponseType(typeof(UserDto), StatusCodes.Status200OK)]
+    public async Task<ActionResult> AddAsync(
+        [FromServices] IRoleRepository roleRepository,
+        [FromBody] SignUpDto input)
+    {
+        try
+        {
+            var role = await roleRepository.GetAsNoTrackingAsync(input.Role.GetHashCode());
+            var user = new User(input.Name, input.Password, role);
+            _userRepository.Insert(user);
+            if (!_unitOfWork.Commit())
```

Comment:
close to the comment below, this uniOfWork behaviour can be moved to an ActionFilter from the mvc pipeline, if it is something that might occur often on every endpoint. 

you can write straightforward controller code: geting, creating and changing entities, returning them as nothing wrong happened, then on an "UnitOfWorkActionFilter" you can check if commit() fails, and override the context response or act as you wish.

If you combine Action filters with exception handler, your code become less repeated, clean and you remove the boilerplate related to formating, errors and etc, as it won't change much and says nothing about business intent.